doc/ADR/cpp_api_modularisation.md

```diff
@@ -0,0 +1,103 @@
+\page cpp-api-and-modularisation Libcprover-cpp and Modularisation
+
+**Date**: 22 Jun 2023
+**Updated**: 22 Jun 2023
+**Author**: Fotis Koutoulakis, fotis.koutoulakis@diffblue.com
+**Domain**: Architecture, API
+**Description**: This document outlines our thinking about the rearchitecting of
+    CBMC using the C++ API (`Libcprover-cpp`) as the central module and the
+    transitioning of other tools to use that as a basis.
+
+## Motivation && Current State of Affairs
+
+CProver is a collection of tools fitting various program analysis needs. It
+has been the product of the evolution of the codebase of the model-checking
+tool for C (`CBMC`), after it has been adopted with various front-ends/back-ends
+and various auxilliary tools.
+
+During this time, the repository has grown organically, using some guidelines
+for development that were based on tradition and intuition rather than some
+tight development guide. This development model has been successful for most
+of CProver's life, based on its nature as a hybrid industrial/academic and
+experimental/applied tool. However, it has had the side-effect of accruing
+some code duplication and technical debt. It has also made the codebase harder
+to understand/develop and integrate with, especially for someone who is new
+to the codebase.
+
+The above concerns have generated discussions about the breaking down of
+CProver into modules, with cleaner interfaces and tighter boundary control
+between them, making the code easier to integrate in other projects (by making
+the various component modules easier to combine and reuse) and making the
+codebase easier to understand and maintain.
+
+## The Plan Going Forward
+
+Given the above outline, we have reached a point where we are strongly motivated
+to take action to better componentise and modularise CProver.
+
+It's also an opportune time for us, given the existence of `Libcprover-cpp`
```

Comment:
Have we defined the architectural design goals for `Libcprover-cpp` anywhere? For example if this is for use in interfacing to the cprover code base from "foreign" languages; are there specific C++ features which should be excluded from use in its "public" headers? I am specifically thinking about the use of language features which may pose calling convention headaches such as exceptions, virtual function calls, overloading; or language features in headers which are understood by C++ compilers, but not C compilers such as templates for example.

Reply:
We do not have such a document in place (yet).